### PR TITLE
Add FileSystem effect

### PR DESCRIPTION
**Pull request by mmhat**
This PR adds a `FileSystem` effect to the `effectful` package as proposed in https://github.com/arybczak/effectful/issues/23#issuecomment-877714896.

### CODE REVIEW

File: effectful/src/Effectful/FileSystem.hs [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+-- | Lifted 'D.withCurrentDirectory'.
+withCurrentDirectory :: FileSystem :> es => FilePath -> Eff es a -> Eff es a
+withCurrentDirectory path k = unsafeEff $ \es -> do
+  seqUnliftEff es $ \unlift ->
+    D.withCurrentDirectory path $ unlift k
+
+----------------------------------------
+-- Pre-defined directories
+
+-- TODO: Should this have the Environment effect, too?
```
Comment (arybczak) [inline effectful/src/Effectful/FileSystem.hs:161]: I don't think so :) Same for all TODOs below.

File: effectful/src/Effectful/FileSystem.hs [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+getCurrentDirectory = unsafeEff_ D.getCurrentDirectory
+
+-- | Lifted 'D.setCurrentDirectory'.
+setCurrentDirectory :: FileSystem :> es => FilePath -> Eff es ()
+setCurrentDirectory = unsafeEff_ . D.setCurrentDirectory
+
+-- | Lifted 'D.withCurrentDirectory'.
+withCurrentDirectory :: FileSystem :> es => FilePath -> Eff es a -> Eff es a
+withCurrentDirectory path k = unsafeEff $ \es -> do
+  seqUnliftEff es $ \unlift ->
```
Comment (arybczak) [inline effectful/src/Effectful/FileSystem.hs:155]: You can use `unsafeLiftMapIO` here, it'll be a bit simpler. I think it's fine to assume that these functions will not use multiple threads. :thinking: Same below.

File: effectful/src/Effectful/FileSystem.hs (hunk 1 above)
Comment (mmhat) [reply effectful/src/Effectful/FileSystem.hs:161]: I don't think so, too :)

File: effectful/src/Effectful/FileSystem.hs (hunk 2 above)
Comment (mmhat) [reply effectful/src/Effectful/FileSystem.hs:155]: Done.

**Comment by mmhat** [comment]
> The (applied) field of effect-tracking systems is still fairly young, especially in the wider Haskell community. I think we also must get comfortable with not getting some things right from the beginning. We will have to learn lessons that time only can teach.

Oh I see! That was the point where I was mistaken: I thought someone out there exists some knowledge about the application of effect systems that I am just not aware of.

Thanks for clarifying; It helped me a lot. :+1: And sorry about the fuss I made by bringing this topic to the table.

**Comment by arybczak** [comment]
Thanks :tada:

**Pull request merged**